### ISSUE TITLE
Kube proxy fails and then hangs without dying.

### ISSUE THREAD

**Issue by krish7919** [kubernetes/kubernetes#48705]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:
Service was not able to communicate with backend deployment.
Checked the logs and found that kube-proxy had panic-ed.
My assumption is it was because the dependencies of kube-proxy did not spin up before kube-proxy itself.

**What you expected to happen**:
Expectation is that kube-proxy would panic in this case and restart/recover.
It recovered after a restart of the kube-proxy on all the agents.

**How to reproduce it (as minimally and precisely as possible)**:
Spin up a new Kubernetes cluster on Azure container Service.
I did it twice and it happened both the times.

**Anything else we need to know?**:
Fix is to restart the kube-proxy. I think it is dependency management related. Stacktrace:

```


docker logs  k8s_kube-proxy_kube-proxy-lhsw2_kube-system_985fbe1c-6193-11e7-af56-000d3a
2666d5_0 
I0705 15:07:02.183142       1 server.go:225] Using iptables Proxier.
I0705 15:07:02.244796       1 server.go:249] Tearing down userspace rules.
I0705 15:07:02.362487       1 conntrack.go:81] Set sysctl 'net/netfilter/nf_conntrack_max' to 131072
I0705 15:07:02.362993       1 conntrack.go:66] Setting conntrack hashsize to 32768
I0705 15:07:02.363239       1 conntrack.go:81] Set sysctl 'net/netfilter/nf_conntrack_tcp_timeout_established' to 86400
I0705 15:07:02.363267       1 conntrack.go:81] Set sysctl 'net/netfilter/nf_conntrack_tcp_timeout_close_wait' to 3600
I0705 15:40:08.124291       1 logs.go:41] http: panic serving 168.63.129.16:58960: runtime error: invalid memory address or nil pointer dereference
goroutine 962 [running]:
net/http.(*conn).serve.func1(0xc4202a4d00)
        /usr/local/go/src/net/http/server.go:1491 +0x12a
panic(0x3a4c580, 0xc420016040)
        /usr/local/go/src/runtime/panic.go:458 +0x243
k8s.io/kubernetes/pkg/proxy/healthcheck.hcHandler.ServeHTTP(0xc4209d28c7, 0x7, 0xc4209d28c0, 0x7, 0xc420103860, 0x6cd0ea0, 0xc420d080
d0, 0xc420fec690)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/proxy/healthcheck/healthcheck.go:197 +0x7c
k8s.io/kubernetes/pkg/proxy/healthcheck.(*hcHandler).ServeHTTP(0xc420c68ff0, 0x6cd0ea0, 0xc420d080d0, 0xc420fec690)
        <autogenerated>:8 +0x8c
net/http.serverHandler.ServeHTTP(0xc42041c080, 0x6cd0ea0, 0xc420d080d0, 0xc420fec690)
        /usr/local/go/src/net/http/server.go:2202 +0x7d
net/http.(*conn).serve(0xc4202a4d00, 0x6cd3fa0, 0xc4208a8400)
        /usr/local/go/src/net/http/server.go:1579 +0x4b7
created by net/http.(*Server).Serve
        /usr/local/go/src/net/http/server.go:2293 +0x44d
```

**Environment**:
- Kubernetes version (use `kubectl version`):
```
kubectl version
Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.0", GitCommit:"d3ada0119e776222f11ec7945e6d860061339aad", GitTreeState:"clean", BuildDate:"2017-06-29T23:15:59Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.6", GitCommit:"7fa1c1756d8bc963f1a389f4a6937dc71f08ada2", GitTreeState:"clean", BuildDate:"2017-06-16T18:21:54Z", GoVersion:"go1.7.6", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration**:
Azure Container Service

- OS (e.g. from /etc/os-release):
```
cat /etc/os-release 
NAME="Ubuntu"
VERSION="16.04.2 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.2 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```

- Kernel (e.g. `uname -a`):
```
 uname -a
Linux k8s-master-DA6EDAE7-0 4.4.0-83-generic #106-Ubuntu SMP Mon Jun 26 17:54:43 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

- Install tools:
- Others:

**Comment by krish7919** [comment]
@kubernetes/sig-network

**Comment by krish7919** [comment]
Ok, so this happened again on Azure Container Service and this affected our live instance used for a demo. And now I am of the opinion that this is not because of a startup dependency problem, but can hit anytime the kube-proxy hits the above error/stacktrace.

**Comment by krish7919** [comment]
Update: I think [this commit](https://github.com/kubernetes/kubernetes/commit/41c4e965c353187889f9b86c3e541b775656dc18) already fixes the issue and azure needs to update to v1.7.

/sig azure

**Comment by caseydavenport** [comment]
@krish7919 can you confirm that this issue is fixed in 1.7? 

If so we can close this.

**Comment by krish7919** [comment]
@caseydavenport The commit above does fix the issue. However, I have not personally tested it as I run on Azure managed Kubernetes and it hasn't updated to v1.7 yet.

Is there any way the Azure team can update us on this?

**Comment by rocketraman** [comment]
Just ran into this problem with Kubernetes 1.6.6 deployed on Azure via ACS-Engine. Obviously, this is a serious issue as it prevents service calls from working.

@brendanburns @colemickens

**Comment by krish7919** [comment]
@rocketraman Sorry for the late comment.
I can consistently reproduce this error when the LB facing service restarts. So, if your public facing Kubernetes service is always up, this should not be seen anymore.

**Comment by feiskyer** [comment]
The problem should be fixed in release-1.7.

@rocketraman @krish7919  Opened a cherry-pick PR to release-1.6 #54839.

**Comment by feiskyer** [comment]
Fixed in #54839, will be included in next 1.6 release.
